### PR TITLE
feat(sast): Parse bql policies and add 11 sast checks

### PR DESCRIPTION
**Pull request by arielkru**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

Note re: the tests (WIP) - _The performance tests are failing since there are no real results for the currently scanned repos, will find new repos that violate the existing rules so the tests run on actual scans, TBD before merge_

## Description

* SAST policies should now be written in BQL syntax
* Added a new sast parser to parse the BQL policies and dump them into a temporary yaml file with the semgrep syntax

## New/Edited policies
* 11 new sast policies that cover all the syntax building blocks

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/sast/checks/base_registry.py [hunk 1]
```diff
@@ -4,21 +4,25 @@
...
 class Registry(BaseCheckRegistry):
     def __init__(self, checks_dir: str) -> None:
         super().__init__(report_type=CheckType.SAST)
-        self.rules: List[str] = []
+        self.rules: List[Dict[str, Any]] = []
         self.checks_dir = checks_dir
         self.logger = logging.getLogger(__name__)
+        self.parser = SastCheckParser()
         self.runner_filter: Optional[RunnerFilter] = None
+        self.temp_semgrep_rules_path = os.path.join(self.checks_dir, 'temp_semgrep_rules.yaml')
```
Comment (bo156) [inline checkov/sast/checks/base_registry.py:25]: I think you should separate the string to a parameter for the init with default value, so it could later be overrided if needed

File: checkov/sast/checks/base_registry.py [hunk 2]
```diff
@@ -79,3 +83,13 @@ def _should_skip_check(self, rule: Dict[str, Any]) -> bool:
...
         return True
+
+    def create_temp_rules_file(self) -> None:
+        rules_obj = {'rules': self.rules}
+        with open(self.temp_semgrep_rules_path, 'w') as tempfile:
+            yaml.safe_dump(rules_obj, tempfile)
+        logging.debug(f'created semgrep temporary rules file at: {self.temp_semgrep_rules_path}')
+
+    def delete_temp_rules_file(self) -> None:
+        os.remove(self.temp_semgrep_rules_path)
```
Comment (bo156) [inline checkov/sast/checks/base_registry.py:94]: I assume this might be called when the file is not created by mistake sometime (from a user or from a bug), so I think you should wrap it in `try/catch FileNotFound`

File: checkov/sast/checks/checks_parser.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+class SastCheckParser:
+    def parse_raw_check_to_semgrep(self, raw_check: Dict[str, Dict[str, Any]]) -> Dict[str, Any]:
+        semgrep_rule: Dict[str, Any] = {}
+        if not self.raw_check_is_valid(raw_check):
+            logging.error(f'cant parse the following policy: {raw_check}')
+        try:
+            semgrep_rule = self.parse_rule_metadata(raw_check, semgrep_rule)
+            semgrep_rule.update(self.parse_definition(raw_check['definition']))
+        except Exception as e:
```
Comment (bo156) [inline checkov/sast/checks/checks_parser.py:21]: what's the point of except here if you just raise it back?
Also, I'm not sure how it works in new python versions (I mainly used 2.7 unfortunately in the past), but I think there is a different behavior between letting the exception be raised or catching it and using `raise e`, as it overrides the stacktrace. Instead you can just use `raise` (without the `e`) and it makes sure to not create a new object/stacktrace

File: checkov/sast/checks/checks_parser.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            semgrep_rule['metadata'][SemgrepAttribute.OWASP.value] = owasp
+
+        return semgrep_rule
+
+    def parse_definition(self, definition: Dict[str, Any]) -> Dict[str, Any]:
+        definitions = force_list(definition)
+        conf = {}
+        if len(definitions) > 1:
+            conf[SemgrepAttribute.PATTERNS.value] = self.get_definitions_list_items(definitions)
+        else:  # since definitions has only one value, we check only definitions[0]
```
Comment (bo156) [inline checkov/sast/checks/checks_parser.py:65]: can it be 0 and then raise an `IndexOutOfBoundsError`?

File: checkov/sast/checks/checks_parser.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            if not isinstance(definition, dict):
+                raise TypeError(f'bad definition type, got {type(definition)} instead of dict')
+            if definition.get(BqlConditionType.OR.value):
+                conf[SemgrepAttribute.PATTERN_EITHER.value] = self.get_definitions_list_items(
+                    definition[BqlConditionType.OR.value])
+            elif definition.get(BqlConditionType.AND.value):
+                conf[SemgrepAttribute.PATTERNS.value] = self.get_definitions_list_items(
+                    definition[BqlConditionType.AND.value])
+            else:
+                # handle a single definition
```
Comment (bo156) [inline checkov/sast/checks/checks_parser.py:76]: perhaps separate this block of code to a new function called `_parse_single_definition`? It is already pretty long.

Overall you can really simplify this function and make it more readable if you separate each big block of "elif" and call it by the condition.
For example on line 93 separate the block to `_parse_condition_with_variable`

File: checkov/sast/record.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+from termcolor import colored
+
+from checkov.common.bridgecrew.severities import Severity
+from checkov.common.models.enums import CheckResult
+from checkov.common.output.record import Record
+from checkov.common.typing import _CheckResult
+
+'''
+        check_class: str,
```
Comment (bo156) [inline checkov/sast/record.py:11]: what is the meaning of this documentation?
Also if you add doc to the entire module it should be before the imports in my opinion

File: checkov/terraform/graph_builder/graph_components/module.py [hunk 7]
```diff
@@ -3,7 +3,7 @@
 import json
 import os
 from copy import deepcopy
-from typing import List, Dict, Any, Set, Callable, Tuple, TYPE_CHECKING
+from typing import List, Dict, Any, Set, Callable, Tuple, TYPE_CHECKING  # noqa: F401
```
Comment (bo156) [inline checkov/terraform/graph_builder/graph_components/module.py:6]: out of interest, why is this needed?

File: tests/sast/registry_tests/test_registry.py [hunk 8]
```diff
@@ -1,24 +1,53 @@
...
+             'severity': 'ERROR', 'languages': ['java'], 'metadata': {
+                'name': 'seam log injection',
+                'cwe': "CWE-95: Improper Neutralization of Directives in Dynamically Evaluated Code ('Eval Injection')",
+                'owasp': 'A03:2021 - Injection'},
+                'patterns': [{'pattern': '$LOG.$INFO($X + $Y,...)'},
+                    {'pattern-either': [{'pattern-inside': 'import org.jboss.seam.log.Log\n...\n'},
+                          {'pattern-inside': 'org.jboss.seam.log.Log $LOG = ...\n...\n'}]},
+                          {'metavariable-regex': {'metavariable': '$INFO',
+                                                  'regex': '(debug|error|fatal|info|trace|warn)'}}]}
 class TestRegistry(unittest.TestCase):
```
Comment (bo156) [inline tests/sast/registry_tests/test_registry.py:25]: It's probably too late now to comment on as this was done previously, but what do you think on stopping usage of `unittest` in SAST and instead use `pytest`?
Currently we have a wierd mix of both in checkov and I believe the intention is to eventually migrate to pytest, so at least when we are writing something new let's do it like that from the get go.

I don't want you to change it now though, just a thought for the future :)

File: tests/sast/test_parser.py [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+            {'metavariable-comparison': {
+                'metavariable': '$KEY',
+                'comparison': '$KEY > 1024'
+                }
+            },
+        ]
+    }
+
+
+def test_complex_policy_parsing():
```
Comment (bo156) [inline tests/sast/test_parser.py:346]: this entire file looks like one big "pytest.mark.parametrize" to me.
however I do like the different test names as it is much more clear like that, but I do think you can name test cases in pytest parametrize - there is a package called "Parametrization" or something like that which supports it I believe.

Again no need to do it now as it is already in progress, but it's good for general knowledfe

**Review by bo156 — CHANGES_REQUESTED** [review]

File: checkov/sast/checks/base_registry.py (hunk 1 above)
Comment (bo156) [reply checkov/sast/checks/base_registry.py:25]: I'm actually going abroad for a week so in order to not block this PR, I changed from "requesting changes" to just "commented", so please still answer these comments :) No need to wait for me to return as long as you handle all of the comments

**Review by bo156 — APPROVED** [review]

File: checkov/sast/checks/base_registry.py (hunk 1 above)
Comment (arielkru) [reply checkov/sast/checks/base_registry.py]: Done

File: checkov/sast/checks/checks_parser.py (hunk 4 above)
Comment (arielkru) [reply checkov/sast/checks/checks_parser.py]: I don't think of an option where a definition is passed empty, but added a condition to make sure its not the case

File: checkov/sast/checks/checks_parser.py (hunk 5 above)
Comment (arielkru) [reply checkov/sast/checks/checks_parser.py]: Great idea

File: checkov/sast/record.py (hunk 6 above)
Comment (arielkru) [reply checkov/sast/record.py]: That's how the original record file is written but I'm not sure what's the my meaning and it really doesnt look necessary, deleting it

**Review by achiar99 — APPROVED** [review]
Nice LGTM

File: checkov/terraform/graph_builder/graph_components/module.py (hunk 7 above)
Comment (arielkru) [reply checkov/terraform/graph_builder/graph_components/module.py]: It raised a mypy failure that was fixed in upstream so this was a patch but actually can be removed

File: tests/sast/test_parser.py (hunk 9 above)
Comment (arielkru) [reply tests/sast/test_parser.py:346]: I just think writing every test separately like this is the most readable way to do it, but if you have another example that looks readable and has other advantages it'd be nice to change a lot of our tests to it

File: checkov/sast/consts.py [hunk 10]
```diff
@@ -12,8 +15,104 @@ def list(cls) -> List[Any]:
...
 SUPPORT_FILE_EXT = {
     SastLanguages.PYTHON: ['py'],
     SastLanguages.JAVA: ['java'],
     SastLanguages.JAVASCRIPT: ['js']
 }
+
+FILE_EXT_TO_SAST_LANG = {
+    'py': SastLanguages.PYTHON,
+    'java': SastLanguages.JAVA,
+    'js': SastLanguages.JAVASCRIPT
```
Comment (gruebel) [inline checkov/sast/consts.py:87]: do we also want to add `'ts': SastLanguages.JAVASCRIPT` or will it be an exclusive language?

**Review by gruebel — COMMENTED** [review]
looks good so far 🍪

File: checkov/sast/consts.py (hunk 10 above)
Comment (arielkru) [reply checkov/sast/consts.py:87]: I'm not sure how ts scanning works with semgrep and if it gets under javascript or not, but Iet's leave it as an open question and possible modification for when we implement the js policies and get the answer
